---
title: "Agua_DADA2 workflow"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

``` {r, echo = T, eval = T}

#BiocManager::install("dada2")

#load packages

library(dada2); packageVersion("dada2")

```


``` {r, echo = T, eval = T}
#set the path to our paired end files
path <- "~/R/urbananoles/AguadillaReads"
list.files(path)

#read in the forward and reverse seqs

fnFs <- sort(list.files(path, pattern = "_R1_001.fastq", full.names = TRUE))
fnRs <- sort(list.files(path, pattern = "_R2_001.fastq", full.names = TRUE))

# Extract the sample names with some string manipulaiton

sample.names <- sapply(strsplit(basename(fnFs), "_"), "[", 1)

# Look at the quality profiles of the forward reads
plotQualityProfile(fnFs[1:19])

# Look at the quality profiles of the Reverse reads
plotQualityProfile(fnRs[1:20])
```




``` {r, echo = T, eval = T}

# place filtered files in the filtered/ directory 
filtFs <- file.path(path, "filtered", paste0(sample.names, "_F_filt.fastq.gz"))
filtRs <- file.path(path, "filtered", paste0(sample.names, "_R_filt.fastq.gz"))
names(filtFs) <- sample.names
names(filtRs) <- sample.names

# Filter and trim reads. Trim based on quality.

out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, truncLen=c(240,160),
              maxN=0, maxEE=c(2,2), truncQ=2, rm.phix=TRUE,
              compress=TRUE, multithread=FALSE) # On Windows set multithread=FALSE

head(out)


# Learn the error rates. These will take like 10 mins to run with this data.

errF <- learnErrors(filtFs, multithread=TRUE)
errR <- learnErrors(filtRs, multithread=TRUE)


# Plot the errors
plotErrors(errF, nominalQ=TRUE)
plotErrors(errR, nominalQ=TRUE)
```


``` {r, echo = T, eval = T}

# Sample inferance

dadaFs <- dada(filtFs, err=errF, multithread=TRUE)
dadaRs <- dada(filtRs, err=errR, multithread=TRUE)

# Inspect what just happened
#dadaFs[[1]]

# Merge paired end reads
mergers <- mergePairs(dadaFs, filtFs, dadaRs, filtRs, verbose=TRUE)

# Contruct the sequence table
seqtab <- makeSequenceTable(mergers)
dim(seqtab)

# Remove chimeras
seqtab.nochim <- removeBimeraDenovo(seqtab, method="consensus", multithread=TRUE, verbose=TRUE)
sum(seqtab.nochim)/sum(seqtab)

# How many reads made it through
getN <- function(x) sum(getUniques(x))
track <- cbind(out, sapply(dadaFs, getN), sapply(dadaRs, getN), sapply(mergers, getN), rowSums(seqtab.nochim))
# If processing a single sample, remove the sapply calls: e.g. replace sapply(dadaFs, getN) with getN(dadaFs)
colnames(track) <- c("input", "filtered", "denoisedF", "denoisedR", "merged", "nonchim")
rownames(track) <- sample.names
track

# Assign taxonomy
taxa <- assignTaxonomy(seqtab.nochim, "~/R/urbananoles/AnoleSeqs/silva_nr99_v138_train_set.fa.gz", multithread=TRUE)

# Lets see those taxa assignments
taxa.print <- taxa # Removing sequence rownames for display only
rownames(taxa.print) <- NULL
head(taxa.print)




```

``` {r, echo = T, eval = T}

library(phyloseq); packageVersion("phyloseq")
library(Biostrings); packageVersion("Biostrings")
library(ggplot2); packageVersion("ggplot2")
library(dbplyr)

# mMake the Metadata file. NEEDS TO BE A REAL DATA.FRAME NOT A TIBBLE. Need to figure out how to import as a dataframe its probably easier than what I did below.

sample.out <- rownames(seqtab.nochim)
title <- c("Anole_Fecal","Anole_Fecal","Anole_Fecal","Anole_Fecal","Anole_Fecal","Anole_Fecal","Anole_Fecal","Anole_Fecal","Anole_Fecal","Anole_Fecal","Anole_Fecal","Anole_Fecal","Anole_Fecal","Anole_Fecal","Anole_Fecal","Anole_Fecal","Anole_Fecal","Anole_Fecal","Anole_Fecal","Anole_Fecal")
loc_name <- c("Aguadilla_Forest","Aguadilla_Forest","Aguadilla_Forest","Aguadilla_Forest","Aguadilla_Forest","Aguadilla_Forest","Aguadilla_Forest","Aguadilla_Forest","Aguadilla_Forest","Aguadilla_Forest","Aguadilla_urban","Aguadilla_urban","Aguadilla_urban","Aguadilla_urban","Aguadilla_urban","Aguadilla_urban","Aguadilla_urban","Aguadilla_urban","Aguadilla_urban","Aguadilla_urban")
env_biome <- c("Natural","Natural","Natural","Natural","Natural","Natural","Natural","Natural","Natural","Natural","Urban","Urban","Urban","Urban","Urban","Urban","Urban","Urban","Urban","Urban")
sample <- c("AF123", "AF124", "AF125", "AF126", "AF127", "AF128", "AF129", "AF130", "AF131", "AF132", "AF133", "AF134", "AF135", "AF136", "AF137", "AF138", "AF139", "AF140", "AF141", "AF142")
Agua <- data.frame(Title = title, Loc_name = loc_name, Env_biome = env_biome, Sample = sample)
rownames(Agua) <- sample.out

# Make the actual phyloseq object!

ps <- phyloseq(otu_table(seqtab.nochim, taxa_are_rows = FALSE), sample_data(Agua), tax_table(taxa))
Agua

# convert sequences into asv names and then add the sequences to another subfield. the summarize our phyloseq object.
dna <- Biostrings::DNAStringSet(taxa_names(ps))
names(dna) <- taxa_names(ps)
ps <- merge_phyloseq(ps, dna)
taxa_names(ps) <- paste0("ASV", seq(ntaxa(ps)))
ps

```

``` {r, echo = T, eval = T}

#some quick plots

plot_richness(ps, x="Loc_name", measures=c("Shannon", "Simpson"), color="Env_biome")


```



















